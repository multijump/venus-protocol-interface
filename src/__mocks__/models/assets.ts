import { Asset } from 'types';

import governanceResponse from '__mocks__/api/governance.json';
import { formatToAsset } from 'clients/api/queries/getAssets';

export const assets: Asset[] = governanceResponse.data.markets
  .map(formatToAsset)
  .filter(asset => !!asset) as Asset[];
