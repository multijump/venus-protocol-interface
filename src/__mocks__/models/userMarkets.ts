import { UserMarket } from 'types';

import { userAssets } from './userAssets';

export const userMarkets: UserMarket[] = [
  {
    id: '1',
    riskLevel: 'MINIMAL',
    name: 'Venus',
    assets: userAssets,
  },
  {
    id: '2',
    riskLevel: 'VERY_HIGH',
    name: 'Gaming',
    assets: userAssets,
  },
  {
    id: '3',
    riskLevel: 'MEDIUM',
    name: 'MetaVerse',
    assets: userAssets,
  },
];
