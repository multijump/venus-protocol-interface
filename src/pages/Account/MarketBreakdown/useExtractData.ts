import BigNumber from 'bignumber.js';
import { useMemo } from 'react';
import { UserMarket } from 'types';
import {
  calculateCollateralValue,
  calculateDailyEarningsCents,
  calculateNetApy,
  calculateYearlyEarningsForAssets,
  convertTokensToWei,
  formatCentsToReadableValue,
  formatToReadablePercentage,
} from 'utilities';

import { SAFE_BORROW_LIMIT_PERCENTAGE } from 'constants/safeBorrowLimitPercentage';

const useExtractData = ({ markets, includeXvs }: { markets: UserMarket[]; includeXvs: boolean }) =>
  useMemo(() => {
    const { totalBorrowCents, totalSupplyCents, borrowLimitCents } = markets.reduce(
      (acc, asset) => ({
        totalBorrowCents: acc.totalBorrowCents.plus(
          asset.borrowBalanceTokens.times(asset.tokenPriceDollars).times(100),
        ),
        totalSupplyCents: acc.totalSupplyCents.plus(
          asset.supplyBalanceTokens.times(asset.tokenPriceDollars).times(100),
        ),
        borrowLimitCents: asset.collateral
          ? acc.borrowLimitCents.plus(
              calculateCollateralValue({
                amountWei: convertTokensToWei({
                  value: asset.supplyBalanceTokens,
                  tokenId: asset.id,
                }),
                tokenId: asset.id,
                tokenPriceDollars: asset.tokenPriceDollars,
                collateralFactor: asset.collateralFactor,
              }).times(100),
            )
          : acc.borrowLimitCents,
      }),
      {
        totalSupplyCents: new BigNumber(0),
        totalBorrowCents: new BigNumber(0),
        borrowLimitCents: new BigNumber(0),
      },
    );

    const yearlyEarningsCents = calculateYearlyEarningsForAssets({
      markets,
      includeXvs,
    });

    const dailyEarningsCentsTmp =
      yearlyEarningsCents && calculateDailyEarningsCents(yearlyEarningsCents).toNumber();

    const netApyPercentageTmp =
      yearlyEarningsCents &&
      calculateNetApy({
        supplyBalanceCents: totalSupplyCents,
        yearlyEarningsCents,
      });

    const safeBorrowLimitCentsTmp = borrowLimitCents.multipliedBy(
      SAFE_BORROW_LIMIT_PERCENTAGE / 100,
    );

    const readableSafeBorrowLimitTmp = formatCentsToReadableValue({
      value: safeBorrowLimitCentsTmp,
    });

    const safeBorrowLimitPercentageTmp = formatToReadablePercentage(
      safeBorrowLimitCentsTmp.multipliedBy(100).dividedBy(borrowLimitCents),
    );

    return {
      dailyEarningsCents: dailyEarningsCentsTmp,
      netApyPercentage: netApyPercentageTmp,
      readableSafeBorrowLimit: readableSafeBorrowLimitTmp,
      safeBorrowLimitPercentage: safeBorrowLimitPercentageTmp,
      totalBorrowCents,
      totalSupplyCents,
      borrowLimitCents,
    };
  }, [JSON.stringify(markets), includeXvs]);

export default useExtractData;
