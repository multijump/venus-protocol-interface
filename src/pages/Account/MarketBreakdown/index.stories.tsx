import { ComponentMeta } from '@storybook/react';
import React from 'react';

import { userAssets } from '__mocks__/models/userAssets';
import { withCenterStory } from 'stories/decorators';
import { PALETTE } from 'theme/MuiThemeProvider/muiTheme';

import MarketBreakdown from '.';

export default {
  title: 'Pages/Account/MarketBreakdown',
  component: MarketBreakdown,
  decorators: [withCenterStory({ width: 1200 })],
  parameters: {
    backgrounds: {
      default: PALETTE.background.default,
    },
  },
} as ComponentMeta<typeof MarketBreakdown>;

export const Default = () => (
  <MarketBreakdown assets={userAssets} riskLevel="MINIMAL" marketName="Venus" includeXvs />
);
