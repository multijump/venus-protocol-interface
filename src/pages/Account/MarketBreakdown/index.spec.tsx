import React from 'react';

import { userMarkets } from '__mocks__/models/userMarkets';
import renderComponent from 'testUtils/renderComponent';

import MarketBreakdown, { MarketBreakdownProps } from '.';
import TEST_IDS from './testIds';

jest.mock('clients/api');

const baseProps: MarketBreakdownProps = {
  market: userMarkets[0],
  includeXvs: true,
};

describe('pages/Account/MarketBreakdown', () => {
  it('renders without crashing', () => {
    renderComponent(<MarketBreakdown {...baseProps} />);
  });

  it.each([true, false])(
    'displays stats and tables correctly when includeXvs is %s',
    includeXvs => {
      const { getByTestId } = renderComponent(
        <MarketBreakdown {...baseProps} includeXvs={includeXvs} />,
      );

      expect(getByTestId(TEST_IDS.stats).textContent).toMatchSnapshot();
      expect(getByTestId(TEST_IDS.tables).textContent).toMatchSnapshot();
    },
  );
});
