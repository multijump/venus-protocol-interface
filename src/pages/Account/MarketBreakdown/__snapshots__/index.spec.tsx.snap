// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pages/Account/MarketBreakdown displays stats and tables correctly when includeXvs is false 1`] = `"Net APY19.8%Daily Earnings$94,105.38Total Supply$173,506,958.02Total Borrow$52,565,813.81Borrow limit used:39.35%Limit:$133,591,754.96.Your safe limit:$106,873,403.97"`;

exports[`pages/Account/MarketBreakdown displays stats and tables correctly when includeXvs is false 2`] = `"Supplied assetsAssetAPY / LTVBalancesorted descendingCollateralBUSD0%80%100,014,675.34132799 BUSDXRP12.61%60%15,734,759.98256005 XRPDOGE0.05%80%11,060,167.25926387 DOGEADA2.66%60%10,318,555.04183194 ADACAKE0.23%60%1,980,031.33044493 CAKEMATIC0.31%60%1,004,137.036265 MATICXVS1.23%60%1,002,618.42733259 XVSBNB94.4%80%172,040.19829215 BNBSXP0%50%117,480.47905137 SXPLTC11.29%60%98,240.80908803 LTCBTCB13.4%80%10.29418859 BTCBTUSD0.01%80%0.104699 TUSDUSDC0%81%0 USDCUSDT0%80%0 USDTETH12.02%80%0 ETHAAVE0.55%55%0 AAVETRX0%60%0 TRXUST3.83e+200%0%0 UST-LUNA0%0%0 LUNA-BUSDAPY / LTV0%80%Balance100,014,675.34132799 BUSDCollateralXRPAPY / LTV12.61%60%Balance15,734,759.98256005 XRPCollateralDOGEAPY / LTV0.05%80%Balance11,060,167.25926387 DOGECollateralADAAPY / LTV2.66%60%Balance10,318,555.04183194 ADACollateralCAKEAPY / LTV0.23%60%Balance1,980,031.33044493 CAKECollateralMATICAPY / LTV0.31%60%Balance1,004,137.036265 MATICCollateralXVSAPY / LTV1.23%60%Balance1,002,618.42733259 XVSCollateralBNBAPY / LTV94.4%80%Balance172,040.19829215 BNBCollateralSXPAPY / LTV0%50%Balance117,480.47905137 SXPCollateralLTCAPY / LTV11.29%60%Balance98,240.80908803 LTCCollateralBTCBAPY / LTV13.4%80%Balance10.29418859 BTCBCollateralTUSDAPY / LTV0.01%80%Balance0.104699 TUSDCollateralUSDCAPY / LTV0%81%Balance0 USDCCollateralUSDTAPY / LTV0%80%Balance0 USDTCollateralETHAPY / LTV12.02%80%Balance0 ETHCollateralAAVEAPY / LTV0.55%55%Balance0 AAVECollateralTRXAPY / LTV0%60%Balance0 TRXCollateralUSTAPY / LTV3.83e+200%0%Balance0 USTCollateral-LUNAAPY / LTV0%0%Balance0 LUNACollateral-Borrowed assetsAssetAPYBalancesorted descending% of limitUSDT-1.44%28,555,475.400227 USDT21.38%BUSD-1.99%13,138,630.30459132 BUSD9.84%ADA-5.02%488,047.80626406 ADA0.16%MATIC-3.53%128,231.44490475 MATIC0.07%BNB-99.45%35,072.83424318 BNB7.11%LTC-11.7%483.92651837 LTC0.02%SXP0.71%120.61448997 SXP0%BTCB-10.09%51.9739429 BTCB0.75%ETH-12.18%22.75859545 ETH0.02%USDC-1.8%10.065486 USDC0%AAVE82.79%0.00000053 AAVE0%XVS-13.62%0 XVS0%XRP-11.55%0 XRP0%DOGE0.83%0 DOGE0%CAKE25374.18%0 CAKE0%TUSD15.64%0 TUSD0%TRX3844038.17%0 TRX0%UST0%0 UST0%LUNA-2.01%0 LUNA0%USDTAPY-1.44%Balance28,555,475.400227 USDT% of limit21.38%BUSDAPY-1.99%Balance13,138,630.30459132 BUSD% of limit9.84%ADAAPY-5.02%Balance488,047.80626406 ADA% of limit0.16%MATICAPY-3.53%Balance128,231.44490475 MATIC% of limit0.07%BNBAPY-99.45%Balance35,072.83424318 BNB% of limit7.11%LTCAPY-11.7%Balance483.92651837 LTC% of limit0.02%SXPAPY0.71%Balance120.61448997 SXP% of limit0%BTCBAPY-10.09%Balance51.9739429 BTCB% of limit0.75%ETHAPY-12.18%Balance22.75859545 ETH% of limit0.02%USDCAPY-1.8%Balance10.065486 USDC% of limit0%AAVEAPY82.79%Balance0.00000053 AAVE% of limit0%XVSAPY-13.62%Balance0 XVS% of limit0%XRPAPY-11.55%Balance0 XRP% of limit0%DOGEAPY0.83%Balance0 DOGE% of limit0%CAKEAPY25374.18%Balance0 CAKE% of limit0%TUSDAPY15.64%Balance0 TUSD% of limit0%TRXAPY3844038.17%Balance0 TRX% of limit0%USTAPY0%Balance0 UST% of limit0%LUNAAPY-2.01%Balance0 LUNA% of limit0%AssetsSuppliedBorrowSort by:Risk level​AssetAPY / LTVBalancesorted descendingCollateralBUSD0%80%100,014,675.34132799 BUSDXRP12.61%60%15,734,759.98256005 XRPDOGE0.05%80%11,060,167.25926387 DOGEADA2.66%60%10,318,555.04183194 ADACAKE0.23%60%1,980,031.33044493 CAKEMATIC0.31%60%1,004,137.036265 MATICXVS1.23%60%1,002,618.42733259 XVSBNB94.4%80%172,040.19829215 BNBSXP0%50%117,480.47905137 SXPLTC11.29%60%98,240.80908803 LTCBTCB13.4%80%10.29418859 BTCBTUSD0.01%80%0.104699 TUSDUSDC0%81%0 USDCUSDT0%80%0 USDTETH12.02%80%0 ETHAAVE0.55%55%0 AAVETRX0%60%0 TRXUST3.83e+200%0%0 UST-LUNA0%0%0 LUNA-BUSDAPY / LTV0%80%Balance100,014,675.34132799 BUSDCollateralXRPAPY / LTV12.61%60%Balance15,734,759.98256005 XRPCollateralDOGEAPY / LTV0.05%80%Balance11,060,167.25926387 DOGECollateralADAAPY / LTV2.66%60%Balance10,318,555.04183194 ADACollateralCAKEAPY / LTV0.23%60%Balance1,980,031.33044493 CAKECollateralMATICAPY / LTV0.31%60%Balance1,004,137.036265 MATICCollateralXVSAPY / LTV1.23%60%Balance1,002,618.42733259 XVSCollateralBNBAPY / LTV94.4%80%Balance172,040.19829215 BNBCollateralSXPAPY / LTV0%50%Balance117,480.47905137 SXPCollateralLTCAPY / LTV11.29%60%Balance98,240.80908803 LTCCollateralBTCBAPY / LTV13.4%80%Balance10.29418859 BTCBCollateralTUSDAPY / LTV0.01%80%Balance0.104699 TUSDCollateralUSDCAPY / LTV0%81%Balance0 USDCCollateralUSDTAPY / LTV0%80%Balance0 USDTCollateralETHAPY / LTV12.02%80%Balance0 ETHCollateralAAVEAPY / LTV0.55%55%Balance0 AAVECollateralTRXAPY / LTV0%60%Balance0 TRXCollateralUSTAPY / LTV3.83e+200%0%Balance0 USTCollateral-LUNAAPY / LTV0%0%Balance0 LUNACollateral-"`;

exports[`pages/Account/MarketBreakdown displays stats and tables correctly when includeXvs is true 1`] = `"Net APY19.8%Daily Earnings$94,115.13Total Supply$173,506,958.02Total Borrow$52,565,813.81Borrow limit used:39.35%Limit:$133,591,754.96.Your safe limit:$106,873,403.97"`;

exports[`pages/Account/MarketBreakdown displays stats and tables correctly when includeXvs is true 2`] = `"Supplied assetsAssetAPY / LTVBalancesorted descendingCollateralBUSD0%80%100,014,675.34132799 BUSDXRP12.61%60%15,734,759.98256005 XRPDOGE0.05%80%11,060,167.25926387 DOGEADA2.66%60%10,318,555.04183194 ADACAKE0.23%60%1,980,031.33044493 CAKEMATIC0.31%60%1,004,137.036265 MATICXVS1.23%60%1,002,618.42733259 XVSBNB94.4%80%172,040.19829215 BNBSXP0%50%117,480.47905137 SXPLTC11.29%60%98,240.80908803 LTCBTCB13.4%80%10.29418859 BTCBTUSD0.01%80%0.104699 TUSDUSDC0%81%0 USDCUSDT0%80%0 USDTETH12.02%80%0 ETHAAVE0.55%55%0 AAVETRX0%60%0 TRXUST3.83e+200%0%0 UST-LUNA0%0%0 LUNA-BUSDAPY / LTV0%80%Balance100,014,675.34132799 BUSDCollateralXRPAPY / LTV12.61%60%Balance15,734,759.98256005 XRPCollateralDOGEAPY / LTV0.05%80%Balance11,060,167.25926387 DOGECollateralADAAPY / LTV2.66%60%Balance10,318,555.04183194 ADACollateralCAKEAPY / LTV0.23%60%Balance1,980,031.33044493 CAKECollateralMATICAPY / LTV0.31%60%Balance1,004,137.036265 MATICCollateralXVSAPY / LTV1.23%60%Balance1,002,618.42733259 XVSCollateralBNBAPY / LTV94.4%80%Balance172,040.19829215 BNBCollateralSXPAPY / LTV0%50%Balance117,480.47905137 SXPCollateralLTCAPY / LTV11.29%60%Balance98,240.80908803 LTCCollateralBTCBAPY / LTV13.4%80%Balance10.29418859 BTCBCollateralTUSDAPY / LTV0.01%80%Balance0.104699 TUSDCollateralUSDCAPY / LTV0%81%Balance0 USDCCollateralUSDTAPY / LTV0%80%Balance0 USDTCollateralETHAPY / LTV12.02%80%Balance0 ETHCollateralAAVEAPY / LTV0.55%55%Balance0 AAVECollateralTRXAPY / LTV0%60%Balance0 TRXCollateralUSTAPY / LTV3.83e+200%0%Balance0 USTCollateral-LUNAAPY / LTV0%0%Balance0 LUNACollateral-Borrowed assetsAssetAPYBalancesorted descending% of limitUSDT-1.44%28,555,475.400227 USDT21.38%BUSD-1.99%13,138,630.30459132 BUSD9.84%ADA-5.02%488,047.80626406 ADA0.16%MATIC-3.53%128,231.44490475 MATIC0.07%BNB-99.45%35,072.83424318 BNB7.11%LTC-11.7%483.92651837 LTC0.02%SXP0.71%120.61448997 SXP0%BTCB-10.09%51.9739429 BTCB0.75%ETH-12.18%22.75859545 ETH0.02%USDC-1.8%10.065486 USDC0%AAVE82.79%0.00000053 AAVE0%XVS-13.62%0 XVS0%XRP-11.55%0 XRP0%DOGE0.83%0 DOGE0%CAKE25374.18%0 CAKE0%TUSD15.64%0 TUSD0%TRX3844038.17%0 TRX0%UST0%0 UST0%LUNA-2.01%0 LUNA0%USDTAPY-1.44%Balance28,555,475.400227 USDT% of limit21.38%BUSDAPY-1.99%Balance13,138,630.30459132 BUSD% of limit9.84%ADAAPY-5.02%Balance488,047.80626406 ADA% of limit0.16%MATICAPY-3.53%Balance128,231.44490475 MATIC% of limit0.07%BNBAPY-99.45%Balance35,072.83424318 BNB% of limit7.11%LTCAPY-11.7%Balance483.92651837 LTC% of limit0.02%SXPAPY0.71%Balance120.61448997 SXP% of limit0%BTCBAPY-10.09%Balance51.9739429 BTCB% of limit0.75%ETHAPY-12.18%Balance22.75859545 ETH% of limit0.02%USDCAPY-1.8%Balance10.065486 USDC% of limit0%AAVEAPY82.79%Balance0.00000053 AAVE% of limit0%XVSAPY-13.62%Balance0 XVS% of limit0%XRPAPY-11.55%Balance0 XRP% of limit0%DOGEAPY0.83%Balance0 DOGE% of limit0%CAKEAPY25374.18%Balance0 CAKE% of limit0%TUSDAPY15.64%Balance0 TUSD% of limit0%TRXAPY3844038.17%Balance0 TRX% of limit0%USTAPY0%Balance0 UST% of limit0%LUNAAPY-2.01%Balance0 LUNA% of limit0%AssetsSuppliedBorrowSort by:Risk level​AssetAPY / LTVBalancesorted descendingCollateralBUSD0%80%100,014,675.34132799 BUSDXRP12.61%60%15,734,759.98256005 XRPDOGE0.05%80%11,060,167.25926387 DOGEADA2.66%60%10,318,555.04183194 ADACAKE0.23%60%1,980,031.33044493 CAKEMATIC0.31%60%1,004,137.036265 MATICXVS1.23%60%1,002,618.42733259 XVSBNB94.4%80%172,040.19829215 BNBSXP0%50%117,480.47905137 SXPLTC11.29%60%98,240.80908803 LTCBTCB13.4%80%10.29418859 BTCBTUSD0.01%80%0.104699 TUSDUSDC0%81%0 USDCUSDT0%80%0 USDTETH12.02%80%0 ETHAAVE0.55%55%0 AAVETRX0%60%0 TRXUST3.83e+200%0%0 UST-LUNA0%0%0 LUNA-BUSDAPY / LTV0%80%Balance100,014,675.34132799 BUSDCollateralXRPAPY / LTV12.61%60%Balance15,734,759.98256005 XRPCollateralDOGEAPY / LTV0.05%80%Balance11,060,167.25926387 DOGECollateralADAAPY / LTV2.66%60%Balance10,318,555.04183194 ADACollateralCAKEAPY / LTV0.23%60%Balance1,980,031.33044493 CAKECollateralMATICAPY / LTV0.31%60%Balance1,004,137.036265 MATICCollateralXVSAPY / LTV1.23%60%Balance1,002,618.42733259 XVSCollateralBNBAPY / LTV94.4%80%Balance172,040.19829215 BNBCollateralSXPAPY / LTV0%50%Balance117,480.47905137 SXPCollateralLTCAPY / LTV11.29%60%Balance98,240.80908803 LTCCollateralBTCBAPY / LTV13.4%80%Balance10.29418859 BTCBCollateralTUSDAPY / LTV0.01%80%Balance0.104699 TUSDCollateralUSDCAPY / LTV0%81%Balance0 USDCCollateralUSDTAPY / LTV0%80%Balance0 USDTCollateralETHAPY / LTV12.02%80%Balance0 ETHCollateralAAVEAPY / LTV0.55%55%Balance0 AAVECollateralTRXAPY / LTV0%60%Balance0 TRXCollateralUSTAPY / LTV3.83e+200%0%Balance0 USTCollateral-LUNAAPY / LTV0%0%Balance0 LUNACollateral-"`;
